src/test/java/com/google/sps/UserServletTest.java

```diff
@@ -0,0 +1,314 @@
+import static com.google.appengine.api.datastore.FetchOptions.Builder.withLimit;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+import com.google.gson.*;
+
+@RunWith(JUnit4.class)
+public final class UserServletTest {
+  private UserServlet servlet;
+  private static final String MOCK_ID_TOKEN = "123";;
+  private static final String MOCK_USER_ID = "00";
+  private static final String MOCK_NICKNAME = "Name";
+
+
+  private static final HashMap<String, Integer> DEFAULT_CHALLENGE_STATUSES = getDefaultChallengeStatuses();
+
+  private final LocalServiceTestHelper helper =
+      new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
+
+  private final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+  private User user;
+  private JSONObject userJson;
+
+  @Before
+  public void setup(){
+    helper.setUp();
+    servlet = new UserServlet();
+    servlet.setIdHelper(new MockIdHelper());
+    servlet.setDatastoreService(datastore);
+
+    user = new User.Builder(MOCK_USER_ID)
+        .setNickname(MOCK_NICKNAME)
+        .setCurrentChallengeId("GARD_0")
+        .setChallengeStatuses(DEFAULT_CHALLENGE_STATUSES)
+        .build();
+    
+    userJson = new JSONObject(user.toJSON());
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  @Test
+  public void getUser(){
+    datastore.put(user.toEntity());
+
+    HttpServletRequest request = mock(HttpServletRequest.class);
+    HttpServletResponse response = mock(HttpServletResponse.class);
+
+    when(request.getParameter(UserServlet.ID_TOKEN_PARAM)).thenReturn(MOCK_ID_TOKEN);
+
+    StringWriter writer = new StringWriter();
+    PrintWriter pw = new PrintWriter(writer);
+
+    try {
+      when(response.getWriter()).thenReturn(pw);
+      servlet.doGet(request, response);
+      pw.flush();
+
+      JSONObject responseJson = new JSONObject(writer.toString().trim());
+      Assert.assertEquals(userJson.getString(User.ID),
+                          responseJson.getString(User.ID));
+      Assert.assertEquals(userJson.getString(User.NICKNAME),
+                          responseJson.getString(User.NICKNAME));
+    } catch (IOException e) {}
+  }
+
+  @Test
+  public void getNonexistentUser(){
+    HttpServletRequest request = mock(HttpServletRequest.class);
+    HttpServletResponse response = mock(HttpServletResponse.class);
+
+    when(request.getParameter(UserServlet.ID_TOKEN_PARAM)).thenReturn(MOCK_ID_TOKEN);
+
+    StringWriter writer = new StringWriter();
+    PrintWriter pw = new PrintWriter(writer);
+
+    try {
+      when(response.getWriter()).thenReturn(pw);
+      servlet.doGet(request, response);
+      pw.flush();
+
+      Assert.assertEquals(writer.toString().trim(), "");
```

Comment:
If possible, you should also assert response.getStatus() == SC_NOT_FOUND

Reply:
Couldn't figure out how to get the status from the response, both success and error returned 0. I'll try to look into it again and see if I was missing something.